tlv/stream.go

```diff
@@ -0,0 +1,259 @@
+// and a set of known records.
+func NewStream(records ...Record) *Stream {
+	// Assert that the ordering of the Records is canonical and appear in
+	// ascending order of type.
+	var min Type
+	var overflow bool
+	for _, record := range records {
+		if overflow || record.typ < min {
+			panic(ErrStreamNotCanonical.Error())
+		}
+		if record.encoder == nil {
+			panic(ErrNoEncoder.Error())
+		}
+		if record.decoder == nil {
+			panic(ErrNoDecoder.Error())
+		}
+		if record.typ == math.MaxUint64 {
+			overflow = true
+		}
+		min = record.typ + 1
+	}
+
+	return &Stream{
+		records: records,
+	}
+}
+
+// Encode writes a Stream to the passed io.Writer. Each of the Records known to
+// the Stream is written in ascending order of their type so as to be canonical.
+//
+// The stream is constructed by concatenating the individual, serialized Records
+// where each record has the following format:
+//    [varint: type]
+//    [varint: length]
+//    [length: value]
+//
+// An error is returned if the io.Writer fails to accept bytes from the
+// encoding, and nothing else. The ordering of the Records is asserted upon the
+// creation of a Stream, and thus the output will be by definition canonical.
+func (s *Stream) Encode(w io.Writer) error {
+	// Iterate through all known records, if any, serializing each record's
+	// type, length and value.
+	for i := range s.records {
+		rec := &s.records[i]
+
+		// Write the record's type as a varint.
+		err := WriteVarInt(w, uint64(rec.typ), &s.buf)
+		if err != nil {
+			return err
+		}
+
+		// Write the record's length as a varint.
+		err = WriteVarInt(w, rec.Size(), &s.buf)
+		if err != nil {
+			return err
+		}
+
+		// Encode the current record's value using the stream's codec.
+		err = rec.encoder(w, rec.value, &s.buf)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Decode deserializes TLV Stream from the passed io.Reader. The Stream will
+// inspect each record that is parsed and check to see if it has a corresponding
+// Record to facilitate deserialization of that field. If the record is unknown,
+// the Stream will discard the record's bytes and proceed to the subsequent
+// record.
+//
+// Each record has the following format:
+//    [varint: type]
+//    [varint: length]
+//    [length: value]
+//
+// A series of (possibly zero) records are concatenated into a stream, this
+// example contains two records:
+//
+//    (t: 0x01, l: 0x04, v: 0xff, 0xff, 0xff, 0xff)
+//    (t: 0x02, l: 0x01, v: 0x01)
+//
+// This method asserts that the byte stream is canonical, namely that each
+// record is unique and that all records are sorted in ascending order. An
+// ErrNotCanonicalStream error is returned if the encoded TLV stream is not.
+//
+// We permit an io.EOF error only when reading the type byte which signals that
+// the last record was read cleanly and we should stop parsing. All other io.EOF
+// or io.ErrUnexpectedEOF errors are returned.
+func (s *Stream) Decode(r io.Reader) error {
```

Comment:
Not a blocker for this PR, but this method would make for a good fuzz testing target. Even before the, we can assert compliance of our encoder by using `testing/quick` here to ensure we can always ingest what we produce, and don't inadvertently create any invalid streams.

Reply:
ooo good idea, maybe we can add it to the fuzzing harnesses that @Crypt-iQ has been working on